clients/src/main/java/org/apache/kafka/common/protocol/MessageUtil.java

```diff
@@ -204,13 +204,17 @@ public static boolean compareRawTaggedFields(List<RawTaggedField> first,
         }
     }
 
-    public static ByteBuffer toByteBuffer(final Message message, final short version) {
+    public static Readable toReadable(final Message message, final short version) {
         ObjectSerializationCache cache = new ObjectSerializationCache();
         int messageSize = message.size(cache, version);
         ByteBufferAccessor bytes = new ByteBufferAccessor(ByteBuffer.allocate(messageSize));
         message.write(bytes, cache, version);
         bytes.flip();
-        return bytes.buffer();
+        return bytes;
+    }
+
+    public static ByteBuffer toByteBuffer(final Message message, final short version) {
+        return ((ByteBufferAccessor) toReadable(message, version)).buffer();
```

Comment:
This is an unsafe cast on the surface. You can avoid it by having a private method that returns `ByteBufferAccessor`.

Reply:
I see a few other unsafe casts in the PR. We should devise a solution that doesn't involve that.

Reply:
I agree, we could probably just have  `ByteAccessor toByteAcessor(final Message message, final short version)` either as a single method to be used for both variants or as a private method with 2 public wrappers `toReadable` and `toByteBuffer`.